File: src/main/java/com/example/Nadeuri/likes/LikeService.java
//package com.example.Nadeuri.likes;
//
//import com.example.Nadeuri.board.BoardEntity;
//import com.example.Nadeuri.board.BoardRepository;
//import com.example.Nadeuri.member.MemberEntity;
//import com.example.Nadeuri.member.MemberRepository;
//import lombok.RequiredArgsConstructor;
//import org.springframework.stereotype.Service;
//import org.springframework.transaction.annotation.Transactional;
//import java.util.Optional;
//
//@Transactional(readOnly = true)
//@RequiredArgsConstructor
//@Service
//public class LikeService {
//    private final LikeRepository likeRepository;
//    private final BoardRepository boardRepository;
//    private final MemberRepository memberRepository;
//
//    /**
//     * 좋아요 등록
//     */
////    @Transactional
////    public void like(final Long boardId, final String userId) {
////        BoardEntity board = retrieveBoard(boardId);
////        MemberEntity member = retrieveMember(userId);
////
////        handleLike(member, board);
////        board.increaseLikeCount();
////    }
//
//    /**
//     * 좋아요 취소
//     */
////    @Transactional
////    public LikeEntity unLike(final Long boardId, final String userId) {
////        BoardEntity board = retrieveBoard(boardId);
////        MemberEntity member = retrieveMember(userId);
////
////        return findLikeByMemberAndBoard(member, board)
////                .map(like -> {
////                    like.setLikeDeleted(true);
////                    board.decreaseLikeCount();
////                    return like;
////                })
////                .orElseThrow(LikeException.LIKE_NOT_EXIST::get);
////    }
////
////    private BoardEntity retrieveBoard(Long boardId) {
////        return boardRepository.findById(boardId)
////                .orElseThrow(BoardException.NOT_FOUND::get);
////    }
//
////    private MemberEntity retrieveMember(String userId) {
////        return memberRepository.findByUserId(userId)
////                .orElseThrow(MemberException.NOT_FOUND::get);
////    }
//    private void handleLike(MemberEntity member, BoardEntity board) {
//        Optional<LikeEntity> existingLike = findLikeByMemberAndBoard(member, board);
//
//        if (existingLike.isPresent()) {
//            restoreLike(existingLike.get());
//        } else {
//            createNewLike(member, board);
//        }
//    }
//
//    private void restoreLike(LikeEntity like) {
//        if (like.isDeleted()) {
//            like.setLikeDeleted(false);
//            likeRepository.save(like);
//        }
//    }
//
//    private void createNewLike(MemberEntity member, BoardEntity board) {
//        LikeEntity newLike = LikeEntity.create(member, board);
//        likeRepository.save(newLike);
//    }
//
//    private Optional<LikeEntity> findLikeByMemberAndBoard(MemberEntity member, BoardEntity board) {
//        return likeRepository.findByMemberAndBoard(member, board);
//    }
//}

File: src/main/java/com/example/Nadeuri/likes/service/LikeService.kt
package com.example.Nadeuri.likes.service

import com.example.Nadeuri.board.kotlin.entity.BoardEntity
import com.example.Nadeuri.board.kotlin.exception.BoardException2
import com.example.Nadeuri.board.kotlin.repository.BoardRepository2
import com.example.Nadeuri.likes.entity.LikeEntity
import com.example.Nadeuri.likes.exception.LikeException
import com.example.Nadeuri.likes.repository.LikeRepository
import com.example.Nadeuri.member.MemberEntity
import com.example.Nadeuri.member.MemberRepository
import com.example.Nadeuri.member.exception.MemberException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Transactional(readOnly = true)
@Service
class LikeService(
    private val likeRepository: LikeRepository,
    private val boardRepository: BoardRepository2,
    private val memberRepository: MemberRepository,
) {

    /**
     * 좋아요 등록
     */
    @Transactional
    fun like(principalName: String, boardId: Long)
    {
        val board = retrieveBoard(boardId)
        val member = retrieveMember(principalName)

        handleLike(member, board)
        board.increaseLikeCount()
    }

    /**
     * 좋아요 취소
     */
    @Transactional
    fun unLike(principalName: String, boardId: Long): LikeEntity
    {
        val board = retrieveBoard(boardId)
        val member = retrieveMember(principalName)

        return findLikeByMemberAndBoard(member, board)?.let { like ->
            like.setLikeDeleted(true)
            board.decreaseLikeCount()
            like
        } ?: throw LikeException.LIKE_NOT_EXIST.get()
    }

    fun handleLike(member: MemberEntity, board: BoardEntity) {
        val existingLike = findLikeByMemberAndBoard(member, board)

        existingLike?.let { like ->
            // existingLike가 있는 경우
            restoreLike(like)
        } ?: run {
            // existingLike가 null일 경우
            createNewLike(member, board)
        }
    }

    fun restoreLike(like: LikeEntity) {
        if (like.isDeleted) {
            like.setLikeDeleted(false)
            likeRepository.save(like)
        }
    }

    fun createNewLike(member: MemberEntity, board: BoardEntity) {
        val newLike = LikeEntity.create(member, board);
        likeRepository.save(newLike);
    }

    fun findLikeByMemberAndBoard(member: MemberEntity, board: BoardEntity): LikeEntity? {
        return likeRepository.findByMemberAndBoard(member, board);
    }

    private fun retrieveMember(userId: String) =
        memberRepository.findByUserId(userId) ?: throw MemberException.NOT_FOUND.get()

    private fun retrieveBoard(boardId: Long) =
        boardRepository.findByIdOrNull(boardId) ?: throw BoardException2.NOT_FOUND.get()
}